subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildTaskGraph.java

```diff
@@ -44,16 +64,44 @@ public void awaitTaskCompletion(Collection<? super Throwable> taskFailures) {
 
     @Override
     public IncludedBuildTaskResource.State getTaskState(BuildIdentifier targetBuild, String taskPath) {
-        return buildControllerFor(targetBuild).getTaskState(taskPath);
```

Comment:
❓Would it make sense to introduce a build controller for the root build instead perhaps?

Reply:
I have tried that. But the current interface and implementation also assume that the controller is responsible for "launching the build" (i.e. doing the task graph calculation) on a separate thread. While the main build is already in the state of doing that on the main thread.

I think we should clean this all up in a follow up. But it would be a bigger refactoring.